lib.nix

```diff
@@ -0,0 +1,131 @@
+      else value;
+
+  # Import a Nickel value produced by the Nixel DSL
+  importFromNickel = mkShell: value:
+    let
+      type = builtins.typeOf value;
+      isNickelDerivation = type: type == "nickelPackage" || type == "nickelShell";
+      importFromNickel_ = importFromNickel mkShell;
+    in
+    if (type == "set") then (
+      let valueType = value.type or ""; in
+      if valueType == "nickelPackage" then
+        derivation (builtins.mapAttrs (_: importFromNickel_) value)
+      else if valueType == "nickelShell" then
+        mkShell (builtins.mapAttrs (_: importFromNickel_) value)
+      else if valueType == "nixDerivation" then
+        (import value.drvPath).${value.outputName or "out"}
+      else if valueType == "nixString" then
+        builtins.concatStringsSep "" (builtins.map importFromNickel_ value.fragments)
+      else
+        builtins.mapAttrs (_: importFromNickel_) value
+      )
+    else if (type == "list") then
+        builtins.map importFromNickel_ value
+    else value;
+
+  # Generate a Nickel program that evaluates the nickel-nix output, passing
+  # the given exported packages, and write it to outFile.
+  computeNickelFile = system: {nickelFile, exportedPkgs}:
+    let
+      exportedJSON = builtins.toFile
+          "inputs.json"
+          (builtins.unsafeDiscardStringContext (builtins.toJSON (exportForNickel exportedPkgs)));
+
+      nickelWithImports = builtins.toFile "eval.ncl" ''
+          let params = {
+            pkgs = import "${exportedJSON}",
+            system = "${system}",
+            nix = import "${./nix.ncl}",
+          } in
+          let nickel_expr | params.nix.NickelExpression = import "${nickelFile}" in
+          nickel_expr.output params
+      '';
+
+    in
+    nickelWithImports;
+
+  # Extract the inputs declared in the Nickel expression.
+  extractInputs = {runCommand, nickel, system}: nickelFile:
+    let
+      fileToCall = builtins.toFile "extract-inputs.ncl" ''
+        let nix = import "${./nix.ncl}" in
+        let nickel_expr | nix.NickelExpression = import "${nickelFile}" in
+        nickel_expr.inputs
+      '';
+      result = runCommand "nickel-inputs.json" {} ''
+        ${nickel}/bin/nickel -f ${fileToCall} export > $out
+      '';
+    in
+    (builtins.fromJSON (builtins.readFile result));
+
+  # Process the inputs declared in the Nickel expression, fetch the corresponding
+  # Nix values, and export them to JSON to be directly usable by the nickel-nix
+  # file
+  exportInputs = system: lib: { declaredInputs, flakeInputs }:
+    let
+      pkgNames = builtins.attrNames declaredInputs;
+      addPackage = name: acc:
+        let inputName = declaredInputs."${name}".input; in
+        if builtins.hasAttr inputName flakeInputs then
+          let
+            input =
+              if inputName == "nixpkgs" then
+                flakeInputs.${inputName}.legacyPackages.${system}
+              else
+                flakeInputs.${inputName}.${system}.packages;
```

Comment:
```suggestion
                flakeInputs.${inputName}.packages.${system};
```

Reply:
Ideally I think this whole magic doesn’t really belong here (it should be up to the user to specify the exact attribute path, like in plain Nix). But we can keep it for a first POC

Reply:
I kinda agree, but for the POC we didn't want to have to specify flake-related stuff that doesn't appear in the Nix version of the shell.